### PR TITLE
Added OAuth2 GET and POST to GraphRBAC.json spec

### PR DESCRIPTION
**Pull request by shanepeckham**
### Purpose of PR

This is needed so that we can programmatically grant OAuth2 consent to an application in Azure Active Directory - we need this in the Python and Go SDKs for Azure. This is the supported and correct way to do this respecting customer approval workflows. Example use case includes Azure Kubernetes Service RBAC integration with Azure Active Directory.

Added to stable. What has been added:

**In paths**:
```
    "/{tenantID}/oauth2PermissionGrants" : {
        "get" : {
         "tags": [
           "OAuth2"
         ],
          "operationId": "OAuth2_Get",
          "produces" : [ "application/json" ],
          "parameters" : [
          {
            "name" : "$filter",
            "in" : "query",
            "required" : false,
            "type" : "string",
            "x-example" : "clientId+eq+'61ed44c3-5a1d-4639-a215-07f25129c6c3'"
          },
          {
            "$ref": "#/parameters/ApiVersionParameter"
          },
          {
            "$ref": "#/parameters/tenantIDInPath"
          }
         ],
          "responses" : {
            "200" : {
              "description" : "OK. The operation was successful.",
              "schema" : {
                "$ref" : "#/definitions/OAuth2"
              }
            }
          }
        },
        "post" : {
         "tags": [
           "OAuth2"
         ],
          "consumes" : [ "application/json" ],
          "operationId": "OAuth2_Post",
          "produces" : [ "application/json" ],
          "parameters" : [ 
            {
            "in" : "body",
            "name" : "body",
            "required" : false,
            "schema" : {
              "$ref" : "#/definitions/OAuth2"
                },
                "x-examples" : {
                    "application/json" : "{\n\t\"odata.type\": \"Microsoft.DirectoryServices.OAuth2PermissionGrant\",\n\t\"clientId\": \"39afbaa2-4a5c-4f5b-9ee3-2c83f09bbc87\", \n\t\"consentType\": \"AllPrincipals\",\n\t\"principalId\": null,\n\t\"resourceId\": \"d3247842-c517-4520-80a7-332690ae2fe4\",\n\t\"scope\": \"user_impersonation\",\n    \"startTime\": \"0001-01-01T00:00:00\",\n    \"expiryTime\": \"9000-01-01T00:00:00\"\n}"
                  }
            },
            {
            "$ref": "#/parameters/ApiVersionParameter"
            },
            {
                "$ref": "#/parameters/tenantIDInPath"
            }
          ],
          "responses" : {
            "201" : {
              "description" : "OK. The operation was successful.",
              "schema" : {
                "$ref" : "#/definitions/OAuth2"
              }
            }
          }
        }
      }
```
**And in definitions**:

```
    "OAuth2" : {
        "properties" : {
          "odata.type" : {
            "type" : "string",
            "description": "Microsoft.DirectoryServices.OAuth2PermissionGrant"
          },
          "clientId" : {
            "type" : "string",
            "description": "The objectId of the Service Principal associated with the app"
          },
          "consentType" : {
            "type" : "string",
            "description": "Typically set to AllPrincipals"
          },
          "principalId" : {
            "type" : "object",
            "description": "Set to null if AllPrincipals is set"
          },
          "resourceId" : {
            "type" : "string",
            "description" : "Service Principal Id of the resource you want to grant"
          },
          "scope" : {
            "type" : "string",
            "description": "Typically set to user_impersonation"
          },
          "startTime" : {
            "type" : "string",
            "description" : "Start time for TTL"
          },
          "expiryTime" : {
            "type" : "string",
            "description" : "Expiry time for TTL"
          }
        },
        "example" : {
          "odata.type" : "odata.type",
          "resourceId" : "resourceId",
          "clientId" : "clientId",
          "scope" : "scope",
          "expiryTime" : "expiryTime",
          "consentType" : "consentType",
          "principalId" : "{}",
          "startTime" : "startTime"
        }
      }
```

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information

- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

**Review by marstr — CHANGES_REQUESTED** [review]
Thanks for being mindful of deployment as you submitted your change.

There are a couple of ways forward:
1. Changes are getting deployed relatively soon: you can just make these changes to the existing stable definition of graphrbac. We'll just throw the DoNotMerge label on, and set a timeline for when it should get merged.
1. If it's going to be a while, and you'd like to publish preview packages in the Go & Python SDKs, move these changes to `specification/graphrbac/data-plane/preview/1.6/graphrbac.json`.
1. It's going to be a while, but there's no use publishing preview packages publicly: Create a branch in this repository by joining the GitHub team azure-rest-api-writers. Instructions on joining that team [here](https://github.com/Azure/azure-rest-api-specs/wiki#branch-strategy). We can update this PR to target that branch and make the changes to the existing stable file. Then I'll review/approve the changes. When the changes are deployed server side, you can submit a new PR from the staging branch to master. Give this PR as context so the automatically assigned reviewer knows they don't need to do any further review.

**Comment by shanepeckham** [comment]
Thanks @marstr , I went for option 1

**Comment by jhendrixMSFT** [comment]
@AutorestCI regenerate azure-sdk-for-go

**Comment by shanepeckham** [comment]
@marstr , the Linter is failing on legacy items not related to my changes, please advise on how you would like me to proceed.

**Comment by jhendrixMSFT** [comment]
@AutorestCI regenerate azure-sdk-for-go

**Comment by shanepeckham** [comment]
A downstream PR for (https://github.com/Azure/azure-cli)[azure cli] has a dependency on this PR:
* https://github.com/Azure/azure-cli/pull/6975

File: specification/graphrbac/data-plane/stable/1.6/graphrbac.json [hunk 1]
```diff
@@ -2985,6 +3054,52 @@
         }
       },
       "description": "Server response for Get tenant domains API call."
+    },
+    "OAuth2" : {
```
Comment (marstr) [inline specification/graphrbac/data-plane/stable/1.6/graphrbac.json]: As per the linter, having this share a name with the Operation Group will create a naming conflict in the generated SDKs. Left unchanged, AutoRest will automatically rename this "OAuth2Model". Given context, I'm not sure )Auth2 is a very expressive name for this anyway. Maybe something like "GrantRequest" or "GrantPermission"?

File: specification/graphrbac/data-plane/stable/1.6/graphrbac.json [hunk 2]
```diff
@@ -1592,6 +1592,75 @@
...
+          "in" : "body",
+          "name" : "body",
+          "required" : false,
+          "schema" : {
+            "$ref" : "#/definitions/OAuth2"
+              },
+              "x-examples" : {
+                  "application/json" : "{\n\t\"odata.type\": \"Microsoft.DirectoryServices.OAuth2PermissionGrant\",\n\t\"clientId\": \"39afbaa2-4a5c-4f5b-9ee3-2c83f09bbc87\", \n\t\"consentType\": \"AllPrincipals\",\n\t\"principalId\": null,\n\t\"resourceId\": \"d3247842-c517-4520-80a7-332690ae2fe4\",\n\t\"scope\": \"user_impersonation\",\n    \"startTime\": \"0001-01-01T00:00:00\",\n    \"expiryTime\": \"9000-01-01T00:00:00\"\n}"
+                }
+          },
```
Comment (marstr) [inline specification/graphrbac/data-plane/stable/1.6/graphrbac.json]: The linter is also complaining about not having a "description" field here.

File: specification/graphrbac/data-plane/stable/1.6/graphrbac.json [hunk 3]
```diff
@@ -1592,6 +1592,75 @@
...
+        "operationId": "OAuth2_Get",
+        "description": "Queries OAuth2 permissions for the relevant SP ObjectId of an app.",
+        "produces" : [ "application/json" ],
+        "parameters" : [
+        {
+          "name" : "$filter",
+          "in" : "query",
+          "required" : false,
+          "type" : "string",
+          "x-example" : "clientId+eq+'61ed44c3-5a1d-4639-a215-07f25129c6c3'"
```
Comment (marstr) [inline specification/graphrbac/data-plane/stable/1.6/graphrbac.json]: The linter is complaining about not having a "description" property here.

**Review by marstr — CHANGES_REQUESTED** [review]
Sorry for the delay on this, @shanepeckham. 

In regards to the new ARM errors, I glanced through and validated that none of them will be required by us. Though, I would personally recommend adding examples to help us validate/test your API.

File: specification/graphrbac/data-plane/stable/1.6/graphrbac.json (hunk 1 above)
Comment (shanepeckham) [reply specification/graphrbac/data-plane/stable/1.6/graphrbac.json]: @marstr , I will change this to Permissions in anticipation that we can add a revoke in the future which does not seem to exist at this stage

File: specification/graphrbac/data-plane/stable/1.6/graphrbac.json (hunk 2 above)
Comment (shanepeckham) [reply specification/graphrbac/data-plane/stable/1.6/graphrbac.json]: Added

File: specification/graphrbac/data-plane/stable/1.6/graphrbac.json (hunk 3 above)
Comment (shanepeckham) [reply specification/graphrbac/data-plane/stable/1.6/graphrbac.json]: Added a description

**Comment by shanepeckham** [comment]
@marstr All requested changes applied

**Review by marstr — CHANGES_REQUESTED** [review]
Other than my one comment, this PR LGTM.

**Comment by shanepeckham** [comment]
@marstr we have another PR dependant on this one which has also been approved https://github.com/Azure/azure-cli/pull/6975

**Comment by shanepeckham** [comment]
@marstr Testing Example for granting permission:
```
{
	"odata.type": "Microsoft.DirectoryServices.OAuth2PermissionGrant",
	"clientId": "a417c97d-2569-4bad-bd0c-bf4b187cad00", // App A - This is the service principal objectId of your app you want to grant permissions to
	"consentType": "AllPrincipals",
	"principalId": null,
	"resourceId": "b8415521-9724-457e-88f7-c97448afe42c", // App B - This is the service principal objectId of the app you want to grant permissions to App A for
	"scope": "user_impersonation",
    "startTime": "0001-01-01T00:00:00",
    "expiryTime": "9000-01-01T00:00:00"
}
```
To test if it worked query:
```
https://graph.windows.net/myorganization/oauth2PermissionGrants?api-version=1.6&$filter=clientId+eq+'App A Service Principal objectId'
```

**Comment by yugangw-msft** [comment]
@marstr can correct me, but @shanepeckham you might need to fix one lint error by change `Permission` to `permission`
```
Property named: 'odata.type', for definition: 'Permissions' must follow camelCase style. Example: 'odata.type'.
```

**Comment by shanepeckham** [comment]
@yugangw-msft , @marstr , the linter error reported (R3016 - DefinitionsPropertiesNamesCamelCase) appears throughout the document on legacy items and to me seems related to the '.' (full stop) in the property name, not the case. Is this correct?

File: specification/graphrbac/data-plane/stable/1.6/graphrbac.json [hunk 4]
```diff
@@ -2985,6 +3056,52 @@
         }
       },
       "description": "Server response for Get tenant domains API call."
+    },
+    "Permissions" : {
+      "properties" : {
+        "odata.type" : {
```
Comment (marstr) [inline specification/graphrbac/data-plane/stable/1.6/graphrbac.json]: I'm not worried about the linter error from this line.

**Review by marstr — CHANGES_REQUESTED** [review]
Clarifying my expectations from the linter :)

Many of the errors that are reported aren't actually new.

File: specification/graphrbac/data-plane/stable/1.6/graphrbac.json [hunk 5]
```diff
@@ -1592,6 +1592,77 @@
...
+      "post" : {
+       "tags": [
+         "OAuth2Permissions_post"
+       ],
+        "consumes" : [ "application/json" ],
+        "operationId": "OAuth2_Post",
+        "description": "Grants OAuth2 permissions for the relevant resource Ids of an app.",
+        "produces" : [ "application/json" ],
+        "parameters" : [ 
+          {
```
[the remark this answers is not in the document]
Comment (shanepeckham) [reply specification/graphrbac/data-plane/stable/1.6/graphrbac.json]: @marstr Added description

**Review by marstr — APPROVED** [review]
The linter errors that are being reported are all now ARM specific requirements, therefor this data-plane library is not beholden to them. (Though for the record, it may help you to add more examples)

**Comment by marstr** [comment]
Let me know when you're ready for me to merge this from the server side. Our expectation is that it is ready for public consumption in at least one Azure region.

**Comment by shanepeckham** [comment]
Thank you @marstr , we are good to go.

**Pull request merged**